### PR TITLE
Fix sorting for objects with positive view z

### PR DESCRIPTION
**Pull request by eiriklegernaes**
**Description**

Currently, if an object with positive z<sub>view</sub> (i.e. center "behind" the camera, but bounds still intersect frustum) is rendered using perspective projection, the computed z<sub>ndc</sub> will be positive, and sorting will not work as intended. This means it'll likely be drawn at the very end, leading to sub-optimal early depth testing and occlusion queries.

This PR changes the sorting input from z<sub>ndc</sub> to -z<sub>view</sub>, which I _think_ will work without breaking existing default and (hopefully) custom sorting.

Thoughts?

sorting_issue

### CODE REVIEW

**Comment by gkjohnson** [comment]
Interesting find. I think if this is going to be adjusted then the best way to do it is to multiply the position vector by the projection matrix _without_ dividing by the `w` component which is what's causing the asymptote and sign flip. If we just sort based the z value in the local camera coordinate frame then the sort order will not be correct if the projection matrix / frustum is rotated - either from use manually setting it or some WebXR cases.

It occurs to me that my recent fix to BatchedMesh sorting in #28401 will have the same issue so we'll want to fix that, as well.

**Comment by eiriklegernaes** [comment]
Makes sense! Thanks for the input, @gkjohnson. I've pushed a change that attempts to implement this. Do you see any obvious problems?

**Comment by mrdoob** [comment]
Seems like the examples that broke all use `OrthographicCamera` and/or `Sprite`.

**Comment by eiriklegernaes** [comment]
Okey. I had to make some silly mistakes, but I believe we're pretty close now. Both the transform from object to world space, and from world to view space are affine transforms, so there's no need to divide by w after doing those.

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1305,7 +1307,8 @@ class WebGLRenderer {
 
 						if ( sortObjects ) {
 
-							_vector3.setFromMatrixPosition( object.matrixWorld )
+							_vector3.setFromMatrixPosition( object.matrixWorld );
+							_vector4.copy( _vector3 )
```
Comment (gkjohnson) [inline src/renderers/WebGLRenderer.js:1311]: I'm wondering if we should add `Vector4.setMatrixFromPosition` or `Vector4.setFromMatrixColumn` so we can facilitate this kind of use more clenaly.

cc @mrdoob

**Comment by WestLangley** [comment]
> If we just sort based the z value in the local camera coordinate frame then the sort order will not be correct if the projection matrix / frustum is rotated - either from use manually setting it or some WebXR cases.

I'd suggest following convention and sorting in camera space. If someone can demonstrate doing so is a problem, we can fix it.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (WestLangley) [reply src/renderers/WebGLRenderer.js:1311]: Let's focus on getting this PR correct, first. If we sort in camera space, this is a non-issue.

Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js]: Can you explain how the previous commits were _not_ correct? It was sorting in the camera clip space before (w divide) which is arguably more correct since this is where the triangle projection and rasterization is actually happening. The fundamental issue is that dividing by w can flip the sign of the object origin.

This is what was done before the commits were overwritten:

> ```js
> _vec3.setFromMatrixPosition( object.matrixWorld );
> _vec4.copy( _vec3 ).applyMatrix4( _projScreenMatrix );
> // sort on _vec4.z ...
> ```

**Comment by gkjohnson** [comment]
> I'd suggest following convention and sorting in camera space. If someone can demonstrate doing so is a problem, we can fix it.

Unless the projection matrix is "facing" exactly along -Z (which there is no restriction for) then this will result in an incorrect sort order.

At JPL our camera models treat +X as boresight which we used to preview a camera view. Even then there is no guarantee that the derived camera models are not rotated or off axis. The [three.js integration with mapbox gl](https://docs.mapbox.com/mapbox-gl-js/example/add-3d-model/) also includes translation and rotation in the projection matrix, as well, which this will break transparent rendering for.

I'll have to check some of my other WebXR camera frustums but my point is that there are real use cases that sorting in camera local space rather than clip space will break. And I'm not sure if I see any value added requiring -Z to be the center direction for custom projection matrices.

**Comment by eiriklegernaes** [comment]
Would it be possible to say that these non-conventional use cases just require custom sorting functions? This will of course break (a tiny portion of?) existing code. 

Let me know which way you want to go on this, and I'll adjust the PR accordingly.

**Comment by gkjohnson** [comment]
> Would it be possible to say that these non-conventional use cases just require custom sorting functions?

There's a reliable, generalized solution to the problem so I'm not sure why custom sorting should be needed. The solution that was implemented previously (referenced in https://github.com/mrdoob/three.js/pull/28474#discussion_r1616742075) should produce correct results, unless I'm missing something, and had passing tests.

> Let me know which way you want to go on this, and I'll adjust the PR accordingly.

It's worth letting the discussion play out before making more changes to the PR. Thank you for the changes!

**Comment by WestLangley** [comment]
tldr;

Sorting in ndc-space was introduced in three.js in 2011. There have been no complaints, but the OP is correct, the sort function is discontinuous, which can lead to incorrect sort order in certain edge cases.

Sorting in camera space is the convention. I am not aware of any engines sorting in clip-space. Does someone know of one?

@gkjohnson You may be correct that sorting in clip-space is better, but I do not like your example. In the mapbox example, the camera projection matrix is a function of the _object's transform_. That seems to me to be a very strange edge case. Also, I am aware of the concept of _skewing_ a projection matrix, but not of the concept of _rotating_ one.

Also to consider, I would expect we want to ensure that the left and right eye render with the same sort order.

**Comment by gkjohnson** [comment]
I'm really surprised there's so much resistance to just doing the math here in the correct space. If we just rely on mathematical consistency we don't have to worry about arbitrarily imposing restrictions. The point is you do not always have control over the construction of the projection matrix - particularly when it comes from another system. The solution I've suggested is correct and the one proposed _only_ serves to break existing use cases because we don't think a user is using it in a way we want. There hasn't really been any other reason given.

I and a lot of other users have likely written a lot of code that this project wouldn't deem the "correct" way of doing things and the fact that it works is great. The flexibility of three.js is why I and many others use it in the first place - the fact that these things work and that users can rely on the mathematics being applied in consistent ways if they know what's going on is a feature, not a bug.

> Also to consider, I would expect we want to ensure that the left and right eye render with the same sort order.

If we were sorting objects twice for VR rendering that would be a reason enough to address this separately for the sake of performance but this isn't happening. Objects from ArrayCameras are sorted once [before they each camera is rendered](https://github.com/mrdoob/three.js/blob/71f8579e56d22d79611eed3b9950a989a9f18ed1/src/renderers/WebGLRenderer.js#L1148-L1152), so this is not an issue.

**Comment by eiriklegernaes** [comment]
While sorting in view space may be the convention, I find it hard to argue against @gkjohnson's points. As far as I can tell, sorting in view and clip space have similar runtime cost and code complexity. But one imposes more restrictions for no apparent benefit.

I was a little quick to force push, but here's the diff for clip space sorting:
https://github.com/eiriklegernaes/three.js/compare/6c8b877c0dd161388...fix-sorting-clip-z

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (gkjohnson) [reply src/renderers/WebGLRenderer.js]: One comment on the code here is that if we don't want to add a new function for vector4 to get the translation column is that this can still be written in one line like so:

```js
_vec4.set( 0, 0, 0, 1 )
  .applyMatrix4( object.matrixWorld )
  .applyMatrix4( _projScreenMatrix ); 
```

Comment (eiriklegernaes) [reply src/renderers/WebGLRenderer.js]: Having a `Vector4.setFromMatrixPosition` makes sense in my mind, at least. If we don't want to add that, I'd say we should avoid the extra vector-matrix multiplication (which due to formatting still adds a line..), and just use the copy function.

**Comment by mrdoob** [comment]
I think we can give this a go.

File: src/renderers/WebGLRenderer.js (hunk 1 above)
Comment (mrdoob) [reply src/renderers/WebGLRenderer.js]: Added `Vector4.setMatrixFromPosition` sounds good to me 👍

**Comment by eiriklegernaes** [comment]
@mrdoob The PR currently contains the view-space sorting. It seems most people now lean towards the clip-space sorting, so I'll have to update this PR prior to merging

**Comment by mrdoob** [comment]
Reverted!
